packages/itwin/measure-tools/src/api/DrawingTypeDataCache.ts

```diff
@@ -9,35 +9,47 @@ import { SheetMeasurementsHelper } from "./SheetMeasurementHelper";
 
 export class DrawingDataCache {
```

Comment:
There's a number of issues I see with this:

1. While rare, viewports can view different imodels, so it's possible that you're updating the cache with an imodel that is not the same imodel the viewport currently knows about.

2. Database ID's are only unique within the imodel, so it's possible to have the same ID for two sheets coming from different imodels.

3. Viewports can be dropped, and you'll have the listener disposers dangling.

I propose:

```ts
export class DrawingDataCache {
  private _modelToSheets: Map<IModelConnection, Map<Id64String, SheetMeasurementsHelper.DrawingTypeData>>;
  private _viewportModelChangedListeners: Map<Viewport, () => void>;

  public constructor() {
    this._modelToSheets = new Map<IModelConnection, Map<Id64String, SheetMeasurementsHelper.DrawingTypeData>>();
    this._viewportModelChangedListeners = new Map<Viewport, () => void>();

    this.setupEvents();
  }

  public getSheetDrawingDataForViewport(vp: Viewport): SheetMeasurementsHelper.DrawingTypeData | undefined {
    if (!vp.view.isSheetView())
      return undefined;

    const cache = this._modelToSheets.get(vp.iModel);
    if (cache)
      return cache.get(vp.view.id);

    return undefined;
  }

  public async querySheetDrawingData(imodel: IModelConnection, sheetId: Id64String): Promise<SheetMeasurementsHelper.DrawingTypeData> {
    let cache = this._modelToSheets.get(imodel);
    if (!cache) {
      cache = new Map<Id64String, SheetMeasurementsHelper.DrawingTypeData>();
      this._modelToSheets.set(imodel, cache);
    }

    let sheetData = cache.get(sheetId);
    if (!sheetData) {
      sheetData = await SheetMeasurementsHelper.getSheetTypes(imodel, sheetId);
      cache.set(sheetId, sheetData);
    }

    return sheetData;
  }

  private setupEvents() {
    // If an imodel closes, clear the cache for it
    IModelConnection.onClose.addListener((imodel) => {
      this._modelToSheets.delete(imodel);
    });

    // Listen for new viewports opening
    IModelApp.viewManager.onViewOpen.addListener((vp) => {
      this.addViewport(vp);
    });

    // Listen for viewports closing, this also is called when IModelApp shuts down
    IModelApp.viewManager.onViewClose.addListener((vp) => {
      this.dropViewport(vp);
    });
  }

  private addViewport(vp: Viewport) {
    vp.onViewedModelsChanged.addListener((viewport) =>{
      if (!viewport.view.isSheetView())
        return;

      void this.querySheetDrawingData(viewport.iModel, viewport.view.id);
    });
  }

  private dropViewport(vp: Viewport) {
    const listener = this._viewportModelChangedListeners.get(vp);
    if (listener) {
      listener();
      this._viewportModelChangedListeners.delete(vp);
    }
  }
}
```

Then you can add this as a property of the `MeasurementManager` for there to be only one instance, that will also drop cached data as imodels close. Something will have to be done about this being publicly constructable, so it may be a good idea to only expose this via a similarly named functions as my get/query methods:

```ts
export class MeasurementManager {
  public async querySheetDrawingData(imodel: IModelConnection, sheetId: Id64String): Promise<SheetMeasurementsHelper.DrawingTypeData>;

  public getSheetDrawingDataForViewport(vp: Viewport): SheetMeasurementsHelper.DrawingTypeData | undefined;
}
```

Reply:
Added initial populate to the code proposal above.

Also I noticed your current code has an array of data per sheet ID, I missed that, but the same idea would still apply.

Reply:
Alright, I've modified to implement almost everything you mentioned.
I think the differences are the drawingTypeData being arrays for the most part and added some singleton logic.